### PR TITLE
Fb 40013 sso auto redirect

### PR DESCRIPTION
**Pull request by labkey-adam**

### CODE REVIEW

File: api/src/org/labkey/api/util/StringUtilsLabKey.java [hunk 1]
```diff
@@ -316,6 +322,54 @@ public static String getDomainName(String name)
...
+            .forEach(list::add);
+
+        difference.entriesDiffering().entrySet().stream()
+            .map(e->e.getKey() + ": " + truncate(e.getValue().leftValue(), truncateLength) + " » " + truncate(e.getValue().rightValue(), truncateLength))
+            .forEach(list::add);
+
+        return String.join(", ", list);
+    }
+
+    private static String truncate(Object o, int truncateLength)
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/util/StringUtilsLabKey.java]: Seems like it belongs in a StringUtils class.  Any reason not to make it public?

File: api/src/org/labkey/api/util/StringUtilsLabKey.java [hunk 2]
```diff
@@ -316,6 +322,54 @@ public static String getDomainName(String name)
...
+     * Compares two maps of name:value pairs and generates a string that documents the entries that have changed
+     * (added, removed, or updated). Useful for audit logging of settings changes. A few examples of output:
+     *
+     *    enabled: true » false
+     *    description: My Configuration » CAS Configuration, enabled: true » false
+     *    serverUrl: » http://localhost:8080/labkey/cas, description: » CAS localhost, autoRedirect: » true
+     *
+     * Null can be passed for either map, in which case an empty map is substituted.
+     */
+    public static @NotNull <V> String getMapDifference(@Nullable Map<String, V> oldMap, @Nullable Map<String, V> newMap, int truncateLength)
```
Comment (labkey-susanh) [inline api/src/org/labkey/api/util/StringUtilsLabKey.java:343]: This is not unlike what @XingY just implemented for displaying detailed audit log changes (GetQueryAuditChangesAction).  There, however, we save the old and new maps in the audit log itself so the client side can do some nice formatting of the differences (see https://github.com/LabKey/sampleManagement/blob/develop/src/client/containers/AuditLog/Components/AuditDetails.tsx).   We're probably going to want to move that client-side code into ui-components at some point and may want to convert the usage of this method in audit events to saving the two maps.  But the utility method is nice, just the same.

**Review by labkey-susanh — APPROVED** [review]

File: api/src/org/labkey/api/util/StringUtilsLabKey.java (hunk 1 above)
Comment (labkey-adam) [reply api/src/org/labkey/api/util/StringUtilsLabKey.java]: okay, done

File: api/src/org/labkey/api/util/StringUtilsLabKey.java (hunk 2 above)
Comment (labkey-adam) [reply api/src/org/labkey/api/util/StringUtilsLabKey.java:343]: Good to know

**Pull request merged**